### PR TITLE
feat: add configuration widget intellisense

### PR DESCRIPTION
**Pull request by matthieu-crouzet**
## Proposed change
add configuration widget intellisense

### CODE REVIEW

File: apps/vscode-extension/src/intellisense/configuration.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+export const configurationCompletionItemProvider: CompletionItemProvider<CompletionItem> = {
+  provideCompletionItems: (doc, pos) => {
+    const completions = Object.entries(configurationTags).map(([label, { detail, description, snippet }]) => {
+      const completion = new CompletionItem({ label: `${label} `, detail }, CompletionItemKind.Keyword);
+      completion.documentation = description;
+      completion.insertText = new SnippetString(`${label} ${snippet}`);
+
+      return completion;
+    });
```
Comment (fpaul-1A) [inline apps/vscode-extension/src/intellisense/configuration.ts:30]: this doesn't seem to rely on the function parameters and could be evaluated only once (moving the code outside the function or using a memoize)
if you need to keep it there, at least you should move the code at the very end of the function, to be evaluated only when it's needed

File: apps/vscode-extension/src/intellisense/configuration.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  },
+  o3rWidgetParam: {
+    description: 'Tag to use CMS widget parameter for configuration',
+    detail: 'paramName paramValue',
+    snippet: '${1:paramName} ${2:paramValue}'
+  }
+};
+
+export const configurationCompletionItemProvider: CompletionItemProvider<CompletionItem> = {
+  provideCompletionItems: (doc, pos) => {
```
Comment (fpaul-1A) [inline apps/vscode-extension/src/intellisense/configuration.ts:23]: this code could use some comments :)

File: apps/vscode-extension/src/intellisense/configuration.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    let prefix = line.slice(0, pos.character);
+    if (!prefix.includes('@')) {
+      return [];
+    }
+
+    const txt = doc.getText();
+    const jsDocCommentMatcher = /\/\*\*[^*](?:\r|\n|.)*?\*\//g;
+    let match = jsDocCommentMatcher.exec(txt);
+    let offset = doc.offsetAt(pos);
+    while (match) {
```
Comment (kpanot) [inline apps/vscode-extension/src/intellisense/configuration.ts:44]: not sure it is actually more performant than a `matchAll` that would easier to read

File: apps/vscode-extension/src/intellisense/configuration.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+const completions = Object.entries(configurationTags).map(([label, { detail, description, snippet }]) => {
+  const completion = new CompletionItem({ label: `${label} `, detail }, CompletionItemKind.Keyword);
+  completion.documentation = description;
+  completion.insertText = new SnippetString(`${label} ${snippet}`);
+
+  return completion;
+});
+
+export const configurationCompletionItemProvider: CompletionItemProvider<CompletionItem> = {
```
Comment (kpanot) [inline apps/vscode-extension/src/intellisense/configuration.ts:32]: would be cleaner to be a factory to initialize your vars:

```suggestion
export const configurationCompletionItemProvider = (): CompletionItemProvider<CompletionItem> => {
  const configurationTags: Record<string, ConfigurationTags> = {
    o3rWidget: {
      description: 'Tag to use CMS widget for configuration',
      detail: 'widgetName',
      // TODO compute snippet thanks to eslint config to propose only valid widgetName
      snippet: '${1:widgetName}'
    },
    o3rWidgetParam: {
      description: 'Tag to use CMS widget parameter for configuration',
      detail: 'paramName paramValue',
      // TODO compute snippet thanks to eslint config to propose only valid paramName and paramValue
      snippet: '${1:paramName} ${2:paramValue}'
    }
  };

  const completions = Object.entries(configurationTags).map(([label, { detail, description, snippet }]) => {
    const completion = new CompletionItem({ label: `${label} `, detail }, CompletionItemKind.Keyword);
    completion.documentation = description;
    completion.insertText = new SnippetString(`${label} ${snippet}`);

    return completion;
  });

  return {
```

File: apps/vscode-extension/src/extension.ts [hunk 5]
```diff
@@ -30,6 +31,7 @@ import {
  */
 export function activate(context: ExtensionContext) {
   context.subscriptions.push(
+    languages.registerCompletionItemProvider(['javascript','typescript'], configurationCompletionItemProvider, '@'),
```
Comment (kpanot) [inline apps/vscode-extension/src/extension.ts:34]: missing readme doc

**Review by kpanot — APPROVED** [review]

**Review by fpaul-1A — APPROVED** [review]

**Pull request merged**